### PR TITLE
[TASK] Reduce node complexity

### PR DESCRIPTION
**Pull request by kitsunet**
This is mostly a cleanup that tries to split the setPath
method into smaller parts. On the way the visibility of some
methods was revised.

### CODE REVIEW

**Comment by kitsunet** [comment]
See #8

**Comment by kitsunet** [comment]
Locally all tests worked now, lets see what travis says.

Important: This is a first step in refactoring. Now the methods are mcuh smaller and cleaner, but what is now left and what is probably the harder task is to find out how the public methods of Node(Interface) that accept a NodeInterface as argument can actually work with any NodeInterface object as currently quite some call internal Node methods and so would fail with any other NodeInterface object. Not that this is a big problem right now, but IMHO that should be cleaned up in a way that allows this.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php [hunk 1]
```diff
@@ -118,6 +118,10 @@ public function getContextPath()
      */
     public function setName($newName)
     {
+        if ($this->getName() === $newName) {
+            return;
+        }
+
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php]: This was moved up from line 129 as it is the fastest check and it actually doesn't matter if newName follows the match pattern when it is already the name of the Node.

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/NodeInterface.php [hunk 2]
```diff
@@ -293,17 +293,6 @@ public function setAccessRoles(array $accessRoles);
...
     /**
-     * Sets the absolute path of this node
-     *
-     * This method is only for internal use by the content repository. Changing
-     * the path of a node manually may lead to unexpected behavior and bad breath.
-     *
-     * @param string $path
-     * @return void
-     */
-    public function setPath($path);
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/NodeInterface.php]: The method was never part of the public API and even states it should not be used. It is purely internal and therefore protected in Node now. That allows removal from the interface which I do not considered breaking.

File: TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php [hunk 3]
```diff
@@ -113,14 +113,14 @@ public function nodeCreationThrowsExceptionIfNodeNameContainsUppercaseCharacters
...
-    public function setPathWorksRecursively()
+    public function setNameWorksRecursively()
     {
         $rootNode = $this->context->getRootNode();
 
         $fooNode = $rootNode->createNode('foo');
         $bazNode = $fooNode->createNode('bar')->createNode('baz');
 
-        $fooNode->setPath('/quux');
+        $fooNode->setName('quux');
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php]: This is the actual change here, unfortunately had to reformat this file, so the other changes are collaterals.

File: TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php [hunk 4]
```diff
@@ -1199,12 +1277,17 @@ public function getPropertyDoesNotReturnNodeReferencesIfTheyAreNotVisibleAccordi
...
+            'e3b99700-f632-4a4c-2f93-0ad07eaf733f'
+        ]);
 
         $nodeB->setHidden(true);
 
         $this->assertNull($nodeA->getProperty('property2'));
-        $this->assertSame(array($nodeC), $nodeA->getProperty('property3'));
+        $property3 = $nodeA->getProperty('property3');
+        $this->assertCount(1, $property3);
+        $this->assertSame($nodeC, reset($property3));
```
Comment (kitsunet) [inline TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php]: This had to be changed as the key numbering is no longer continuous due to the use of array_filter. This was never publicised as feature or expectation and won't be noticed in loops, just a direct comparison of the array with an expectation reveals it, therefore I adapted the assertions here.

**Comment by kitsunet** [comment]
:+1:

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php [hunk 5]
```diff
@@ -761,43 +877,52 @@ public function hasProperty($propertyName)
...
+     *
+     * @param boolean $returnNodesAsIdentifiers
+     * @param array $value
+     * @return array
+     */
+    protected function resolvePropertyReferences($returnNodesAsIdentifiers, $value = [])
+    {
+        if ($returnNodesAsIdentifiers) {
+            return $value;
         }
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php]: why not do that logic in ``getProperty`` like it's done with the ``reference`` type? seems odd sending it as a argument just to return the next argument if it's true..

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php [hunk 6]
```diff
@@ -148,72 +148,187 @@ public function setName($newName)
...
+        $this->moveNodeToDestinationPath($nodeToMove, $variantDestinationPath);
+
+        return [$nodeToMove, $variantOriginalPath, $nodeToMove->getPath()];
+    }
+
+    /**
+     * Moves the given node to the destination path by modifying the underlaying NodeData object.
+     *
+     * @param NodeInterface $node
+     * @param string $destinationPath
```
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php]: @return void

**Comment by aertmann** [comment]
Hey great change @kitsunet! Thanks for cleaning that stuff up, was certainly a mess. Found two minor nitpicks, aside from that it should be good to go from my side.

File: TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php (hunk 4 above)
Comment (aertmann) [inline TYPO3.TYPO3CR/Tests/Functional/Domain/NodesTest.php]: I agree

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/NodeInterface.php (hunk 2 above)
Comment (aertmann) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/NodeInterface.php]: I agree

**Comment by kitsunet** [comment]
\o/ thanks a lot, will fix ASAP :)

File: TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php (hunk 5 above)
Comment (kitsunet) [inline TYPO3.TYPO3CR/Classes/TYPO3/TYPO3CR/Domain/Model/Node.php]: Yep, fair enough. Changed.

**Comment by kitsunet** [comment]
ok, fixed and updated to master. :+1: from me, lets get this done

**Comment by aertmann** [comment]
:+1: by reading

**Pull request merged**